File: 2024/Golang/day1/part2.go
package day1

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"Golang/utils"
)

func Part2() {
    var array1 []int
	map1 := make(map[int]int)
	file, err := os.Open("../Input/day1.1.txt")
	utils.Check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		temp, err := strconv.Atoi(strings.Split(line, " ")[0])
		utils.Check(err)
		array1 = append(array1, temp)
		temp, err = strconv.Atoi(strings.Split(line, " ")[3])
		utils.Check(err)
		map1[temp]++
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
	}

	sum := 0
	for i := 0; i < len(array1); i++ {
		if _, exists := map1[array1[i]]; exists {
			sum += array1[i] * map1[array1[i]]
		}
	}
	fmt.Println(sum)
}

File: 2024/Rust/day1/part2.rs
use crate::utils;
use std::collections::HashMap;

pub fn run_part2() {
    let mut array1: Vec<i32> = Vec::new();
    let mut map: HashMap<i32, i32> = HashMap::new();


    for line in utils::read_lines("../Input/day1.1.txt") {
        let parts: Vec<&str> = line.split(' ').collect();
        if parts.len() > 3 {
            array1.push(parts[0].parse::<i32>().unwrap());
            let key = parts[3].parse::<i32>().unwrap();
            if !map.contains_key(&key) {
                map.insert(key, 1);
            } else {
                let count = map.get(&key).unwrap();
                map.insert(key, count + 1);
            }
        }
    }

    let mut sum = 0;
    for a in array1.iter() {
        if map.contains_key(&a) {
            sum += a * map.get(&a).unwrap();
        }
    }
    println!("{}", sum);
}